### PR TITLE
Extend io customization support: whole obj rule in split obj.

### LINKED ISSUE #8428: I/O customization rule not run on split sub-object of a non-collection object.
This is seen in https://cdcvs.fnal.gov/redmine/issues/25893 and is triggered by a change in CLHEP::Hep3Vector (from 3 individual members to a single array) that is stored in a sub-object of a simple object (no collection involved, it works in that case).

### PR DESCRIPTION
**Pull request by pcanal**
Extend the support for a rule that applies to several data members (and thus currently is applied as the 'object level') to
the case of a split object embedded within a split subobject of an object (previous it worked only for the collection case).

This fixes issue #8428 , see also https://cdcvs.fnal.gov/redmine/issues/25893.

This PR fixes #8428

### CODE REVIEW

**Comment by pcanal** [comment]
test pending.

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by chissg** [comment]
Back-ported to [root-project:v6-22-00-patches](https://github.com/root-project/root/tree/v6-22-00-patches) at [chissg:hotfix/v6-22-00-patches-8428](https://github.com/chissg/root/tree/hotfix/v6-22-00-patches-8428). Local build successful on `scisoftbuild01`, fix validation in progress by Kyle.

**Review by jblomer — COMMENTED** [review]
Looks fine to me although I don't understand all the implications.  Perhaps a comment could be added on the meaning of `fElemId != (UInt_t)-1`.

File: tree/tree/src/TBranchElement.cxx [hunk 1]
```diff
@@ -3541,7 +3591,8 @@ void TBranchElement::InitializeOffsets()
             // to remove it's name to get the correct real data offsets
             ////////////////////////////////////////////////////////////////////
 
-            if( dynamic_cast<TBranchSTL*>(fParent) && stlParentName.Length() )
+            const bool isToplevelCollection = (this == GetMother() && (fType == 3 || fType == 4));
+            if( (dynamic_cast<TBranchSTL*>(fParent) || isToplevelCollection) && stlParentName.Length() )
```
Comment (jblomer) [inline tree/tree/src/TBranchElement.cxx:3595]: Is this related to the rest of the PR?

**Review by jblomer — CHANGES_REQUESTED** [review]
It would be helpful to have a test for the fix, maybe using `CLHEP::Hep3Vector`.

I'd have some weakly related comments questions:
  - Apparently `TBranchElement::fType` can be -2, which is missing from the member documentation
  - Similarly `TBranchElement::fID` can be -2, which seems like being a special value worth documenting
  - The order of serial numbers seems to have a meaning (`fID` <= or >= than `other.fID`); what is this meaning?

**Comment by Axel-Naumann** [comment]
@phsft-bot build

**Comment by pcanal** [comment]
@Axel-Naumann ping (in Jakob's absence)

**Comment by pcanal** [comment]
@phsft-bot build

File: tree/tree/src/TBranchElement.cxx [hunk 2]
```diff
@@ -2058,14 +2058,28 @@ static void GatherArtificialElements(const TObjArray &branches, TStreamerInfoAct
...
+         if (prefix.Length() && nextel->IsA() == TStreamerBase::Class()) {
+            // We skip the name of the base class if there is already a prefix.
+            // See TBranchElement::Unroll
+            subprefix = prefix;
+         } else {
+            subprefix = ename + ".";
+         }
+         auto nbranches = search->GetEntriesFast();
+         for (Int_t bi = 0; bi < nbranches; ++bi) {
+            TBranchElement* subbe = (TBranchElement*)search->At(bi);
```
Comment (Axel-Naumann) [inline tree/tree/src/TBranchElement.cxx:2072]: `search` can be `be->GetListOfBranches()`. Are we guaranteed that `nbranches` is the size for that one, too?

File: tree/tree/src/TBranchElement.cxx [hunk 3]
```diff
@@ -2235,7 +2241,7 @@ void TBranchElement::InitInfo()
                if (parent && parent->fOnfileObject == nullptr)
                   parent->fOnfileObject = fOnfileObject;
             }
-            // Propage this to all the other branch of this type.
+            // Propagate this to all the other branch belonging to the same object.
```
Comment (Axel-Naumann) [inline tree/tree/src/TBranchElement.cxx:2244]:
```suggestion
            // Propagate this to all the other branches belonging to the same object.
```

File: tree/tree/src/TBranchElement.cxx [hunk 4]
```diff
@@ -2248,7 +2254,60 @@ void TBranchElement::InitInfo()
                currentVersion = info->GetClassVersion();
             }
 
-            for (Int_t i = 0; i < nbranches; ++i) {
+            // First find the first branch corresponding to the same class as 'this'
+            // branch
+            Int_t index = branches->IndexOf(this);
+            Int_t firstindex = -1;
+            Int_t lastindex = nbranches - 1;
+            if (index > 0) {
```
Comment (Axel-Naumann) [inline tree/tree/src/TBranchElement.cxx:2262]: Can `index <= 0` happen? Is this worth a `Warning`?

File: tree/tree/src/TBranchElement.cxx [hunk 5]
```diff
@@ -2248,7 +2254,60 @@ void TBranchElement::InitInfo()
...
+                     }
+                     if (!subbranch->fInfo)
+                        subbranch->SetupInfo();
+                     firstindex = i;
+                  }
+                  for(Int_t i = index; i < nbranches; ++i) {
+                     TBranchElement* subbranch = (TBranchElement*)branches->At(i);
+                     TString subbranch_name(subbranch->GetFullName());
+                     if ( ! subbranch_name.BeginsWith(thisprefix)) {
+                        lastindex = i - 1;
```
Comment (Axel-Naumann) [inline tree/tree/src/TBranchElement.cxx:2289]: Can you make this and the setting of `firstindex` (currently `firstindex = i;`) symmetrical, i.e. either update both during all iterations, or set it only once before the `break`?

File: tree/tree/src/TBranchElement.cxx [hunk 6]
```diff
@@ -2416,7 +2475,17 @@ void TBranchElement::InitInfo()
                localInfo = FindOnfileInfo(fClonesClass, fBranches);
             }
 
-            TString prefix(GetName());
+            TString prefix(GetFullName());
+            if (fType == 2 && fID >= 0) {
+               auto start = prefix.Length();
+               if (prefix[prefix.Length()-1] == '.')
```
Comment (Axel-Naumann) [inline tree/tree/src/TBranchElement.cxx:2481]:
```suggestion
               if (prefix[start-1] == '.')
```
(Possibly more obvious given the `--start` right after)

File: tree/tree/src/TBranchElement.cxx [hunk 7]
```diff
@@ -3541,7 +3610,8 @@ void TBranchElement::InitializeOffsets()
             // to remove it's name to get the correct real data offsets
             ////////////////////////////////////////////////////////////////////
 
-            if( dynamic_cast<TBranchSTL*>(fParent) && stlParentName.Length() )
+            const bool isToplevelCollection = (this == GetMother() && (fType == 3 || fType == 4));
+            if( (dynamic_cast<TBranchSTL*>(fParent) || isToplevelCollection) && stlParentName.Length() )
```
Comment (Axel-Naumann) [inline tree/tree/src/TBranchElement.cxx:3614]:
```suggestion
            if( stlParentName.Length() && (isToplevelCollection || dynamic_cast<TBranchSTL*>(fParent)) )
```
saves CPU cycles

**Review by Axel-Naumann — APPROVED** [review]

File: tree/tree/src/TBranchElement.cxx (hunk 2 above)
Comment (pcanal) [reply tree/tree/src/TBranchElement.cxx:2072]: I am confused, isn't `nbranches` set 2 lines above from `search`?

File: tree/tree/src/TBranchElement.cxx (hunk 6 above)
Comment (pcanal) [reply tree/tree/src/TBranchElement.cxx:2481]: ok.

File: tree/tree/src/TBranchElement.cxx (hunk 7 above)
Comment (pcanal) [reply tree/tree/src/TBranchElement.cxx:3614]: nice.

File: tree/tree/src/TBranchElement.cxx (hunk 4 above)
Comment (pcanal) [reply tree/tree/src/TBranchElement.cxx:2262]: Yes, all the time.  It is the case of a toplevel branch.

File: tree/tree/src/TBranchElement.cxx (hunk 5 above)
Comment (pcanal) [reply tree/tree/src/TBranchElement.cxx:2289]: Fair enough.

File: tree/tree/src/TBranchElement.cxx (hunk 4 above)
Comment (pcanal) [reply tree/tree/src/TBranchElement.cxx:2262]: But there was a problem!  The case `index == 0` was incorrectly handled.
```suggestion
            if (index >= 0) {
```
